File: src/code-samples/langchain/middleware-tool-call-monitoring.ts
// :snippet-start: middleware-tool-call-monitoring-js
import { createMiddleware } from "langchain";

const toolMonitoringMiddleware = createMiddleware({
  name: "ToolMonitoringMiddleware",
  wrapToolCall: (request, handler) => {
    console.log(`Executing tool: ${request.toolCall.name}`);
    console.log(`Arguments: ${JSON.stringify(request.toolCall.args)}`);
    try {
      const result = handler(request);
      console.log("Tool completed successfully");
      return result;
    } catch (e) {
      console.log(`Tool failed: ${e}`);
      throw e;
    }
  },
});
// :snippet-end:

// :remove-start:
async function main() {
  if (!toolMonitoringMiddleware) {
    throw new Error("toolMonitoringMiddleware should be defined");
  }
  console.log("✓ Tool call monitoring middleware definition is valid");
}
main();
// :remove-end:

File: src/code-samples/langchain/middleware-tool-call-monitoring.py
# :snippet-start: middleware-tool-call-monitoring-decorator-py
from langchain.agents.middleware import wrap_tool_call
from langchain.tools.tool_node import ToolCallRequest
from langchain.messages import ToolMessage
from langgraph.types import Command
from typing import Callable


@wrap_tool_call
def monitor_tool(
    request: ToolCallRequest,
    handler: Callable[[ToolCallRequest], ToolMessage | Command],
) -> ToolMessage | Command:
    print(f"Executing tool: {request.tool_call['name']}")
    print(f"Arguments: {request.tool_call['args']}")
    try:
        result = handler(request)
        print(f"Tool completed successfully")
        return result
    except Exception as e:
        print(f"Tool failed: {e}")
        raise


# :snippet-end:

# :snippet-start: middleware-tool-call-monitoring-class-py
from langchain.tools.tool_node import ToolCallRequest
from langchain.agents.middleware import AgentMiddleware
from langchain.messages import ToolMessage
from langgraph.types import Command
from typing import Callable

class ToolMonitoringMiddleware(AgentMiddleware):
    def wrap_tool_call(
        self,
        request: ToolCallRequest,
        handler: Callable[[ToolCallRequest], ToolMessage | Command],
    ) -> ToolMessage | Command:
        print(f"Executing tool: {request.tool_call['name']}")
        print(f"Arguments: {request.tool_call['args']}")
        try:
            result = handler(request)
            print(f"Tool completed successfully")
            return result
        except Exception as e:
            print(f"Tool failed: {e}")
            raise


# :snippet-end:

# :remove-start:
if __name__ == "__main__":
    assert monitor_tool is not None, "Decorator middleware should be defined"
    print("✓ Tool call monitoring middleware definitions are valid")
# :remove-end:
